server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1507,6 +1510,99 @@ public void testSnapshotFromStreamInput() throws IOException {
         assertEquals(ops, readOperations);
     }
 
+    public void testSnapshotTrimmedOperations() throws Exception {
+        List<Translog.Index> operations = new ArrayList<>();
+        int translogOperations = randomIntBetween(10, 100);
+        int maxTrimmedSeqNo = translogOperations - randomIntBetween(4, 8);
+
+        final long initialPrimaryTerm = this.primaryTerm.get();
+        for (int op = 0; op < translogOperations; op++) {
+            String ascii = randomAlphaOfLengthBetween(1, 50);
+            Translog.Index operation = new Translog.Index("test", "" + op, op,
+                initialPrimaryTerm, ascii.getBytes("UTF-8"));
+            operations.add(operation);
+        }
+        Randomness.shuffle(operations);
+
+        for (Translog.Index operation : operations) {
+            translog.add(operation);
+        }
+
+        primaryTerm.incrementAndGet();
+        translog.rollGeneration();
+
+        translog.trimOperations(primaryTerm.get(), maxTrimmedSeqNo);
+        translog.sync();
+
+        Set<Long> expectedSeqNo = LongStream.range(0, maxTrimmedSeqNo + 1).boxed().collect(Collectors.toSet());
+        try (Translog.Snapshot snapshot = translog.newSnapshot()) {
+            Translog.Operation next;
+            while ((next = snapshot.next()) != null) {
+                assertThat(expectedSeqNo.remove(next.seqNo()), is(true));
+            }
+            assertThat(snapshot.skippedOperations(), is(translogOperations - (maxTrimmedSeqNo + 1)));
+        }
+        assertThat(expectedSeqNo, hasSize(0));
+    }
+
+    public void testRandomExceptionsOnTrimOperations( ) throws Exception {
+        Path tempDir = createTempDir();
+        final FailSwitch fail = new FailSwitch();
+        fail.failNever();
+        TranslogConfig config = getTranslogConfig(tempDir);
+        List<FileChannel> fileChannels = new ArrayList<>();
+        final Translog failableTLog =
+            getFailableTranslog(fail, config, randomBoolean(), false, null, createTranslogDeletionPolicy(), fileChannels);
+
+        expectThrows(IOException.class,
```

Comment:
This enforce that the exception is thrown, but the current code doesn't *always* throw one.

Reply:
ping!